src/smp.rs

```diff
@@ -54,15 +62,45 @@ impl PerCpu {
         VmIdTracker::init();
 
         // Now initialize each PerCpu structure.
+        // Unwrap okay: PER_CPU_PAGES is non zero.
+        let mut pcpu_stacks =
+            pcpu_pages.into_chunks_iter(core::num::NonZeroU64::new(PER_CPU_PAGES).unwrap());
         for i in 0..cpu_info.num_cpus() {
             let cpu_id = CpuId::new(i);
+            // Unwrap okay. We allocated the area `with num_cpus() * PER_CPU_PAGES` pages.
+            let pcpu_stack = pcpu_stacks.next().unwrap();
+            // Boot CPU is special. Doesn't use the PCPU_BASE area as stack, only for the PCPU area.
+            // TODO: Do not allocate PCPU area for boot cpu.
+            let stack_pages = if cpu_id == boot_cpu {
+                Self::boot_cpu_stack()
+            } else {
+                // Change state from InternalClean to InternalDirty. Pages are clean but this is not
+                // important for the stack (in the boot CPU case, pages are dirty because the stack is
+                // in use).
+                // Safe because the chunk is composed by hypervisor pages owned by `pcpu_stack`.
+                unsafe {
+                    SequentialPages::<InternalDirty>::from_mem_range(
+                        pcpu_stack.base(),
+                        pcpu_stack.page_size(),
+                        pcpu_stack.len(),
+                    )
+                    .unwrap()
+                }
+            };
+            let stack_top = if cpu_id == boot_cpu {
+                crate::hyp_layout::HYP_STACKTOP
+            } else {
+                // Secondary CPUs have PerCpu structure at top of stack.
+                crate::hyp_layout::HYP_STACKTOP - core::mem::size_of::<PerCpu>() as u64
```

Comment:
so by keeping `PerCpu` on the stack, we effectively make it impossible for one CPU to access another's `PerCpu`. we don't use it for inter-processor communication currently, but it's not unreasonable that we would want to do so in the future. have you considered putting the `PerCpu`s in a global array instead? i only put them on the stack to begin with because it was convenient to be able to have one pointer for both the initial sp and `PerCpu` when launching a secondary CPU.

Reply:
So the trick here is that only the CPU using the stack doesn't have its stack mapped in the 1:1. Which is why the CPU bootstrap can wait for the CPU to go online in `start_secondary_cpus` -- which effectively requires access of other CPU's PerCpu from the bootstrap CPU.

My comment above (about removing the need for a full stack for the Bootstrap PerCpu Data structure) implies exactly what you suggest. I had implemented it, but I wanted this patch to be as simple as possible and use the current SMP architecture to review the mechanism of stack allocation.

Reply:
PerCpu not on stack anymore.